build/commands/scripts/sync.js

```diff
@@ -41,40 +49,189 @@ const maybeInstallDepotTools = (options = config.defaultOptions) => {
+                                                                '%False%'
+      }
+    },
+    {
+      managed: '%False%',
+      name: 'src/brave',
+      // We do not use gclient to manage brave-core, so this should not
+      // actually get used.
+      url: 'https://github.com/brave/brave-core.git'
+    }
+  ])
+
+  if (process.env.GIT_CACHE_PATH) {
+    out += toGClientConfigItem('cache_dir', process.env.GIT_CACHE_PATH)
+  }
+  if (config.targetOS) {
+    out += toGClientConfigItem('target_os', [config.targetOS], false)
+  }
+
+  fs.writeFileSync(config.defaultGClientFile, out)
+}
+
+function shouldUpdateChromium(latestSuccessfulSync, expectedSuccessfulSync) {
+  const chromiumRef = expectedSuccessfulSync.chromiumRef
+  const headSHA = util.runGit(config.srcDir, ['rev-parse', 'HEAD'], true)
+  const targetSHA = util.runGit(config.srcDir, ['rev-parse', chromiumRef], true)
+  const needsUpdate = targetSHA !== headSHA || (!headSHA && !targetSHA) ||
+      JSON.stringify(latestSuccessfulSync) !==
+          JSON.stringify(expectedSuccessfulSync)
+  if (needsUpdate) {
+    const currentRef = util.getGitReadableLocalRef(config.srcDir)
+    console.log(
+        `Chromium repo ${chalk.blue.bold('needs sync')}.\n  target is ${
+            chalk.italic(chromiumRef)} at commit ${
+            targetSHA || '[missing]'}\n  current commit is ${
+            chalk.italic(currentRef || '[unknown]')} at commit ${
+            chalk.inverse(
+                headSHA || '[missing]')}\n  latest successful sync is ${
+                  JSON.stringify(latestSuccessfulSync, null, 4)}`)
+  } else {
+    console.log(
+        chalk.green.bold(`Chromium repo does not need sync as it is already ${
+            chalk.italic(
+                chromiumRef)} at commit ${targetSHA || '[missing]'}.`))
+  }
+  return needsUpdate
+}
+
+function syncChromium(program) {
+  const requiredChromiumRef = config.getProjectRef('chrome')
+  let args = [
+    'sync', '--nohooks', '--reset', '--revision',
+    'src@' + requiredChromiumRef, '--with_tags',
+    '--with_branch_heads', '--upstream'
+  ];
+
+  const syncWithForce = program.init || program.force
+  if (syncWithForce) {
+    args.push('--force')
+  }
+  if (program.cleanup) {
+    if (util.isGitExclusionExists(config.srcDir, 'brave/')) {
+      args.push('-D')
+    } else {
+      Log.warn(
+          '--cleanup is ignored, because brave dir is not excluded from src')
+    }
+  }
+
+  const latestSuccessfulSyncFile =
+      path.join(config.rootDir, '.brave_latest_successful_sync.json')
+  const latestSuccessfulSync = util.readJSON(latestSuccessfulSyncFile)
+  const expectedSuccessfulSync = {
+    chromiumRef: requiredChromiumRef,
+    gClientTimestamp: fs.statSync(config.gClientFile).mtimeMs.toString()
+  }
+
+  const chromiumNeedsUpdate =
+      shouldUpdateChromium(latestSuccessfulSync, expectedSuccessfulSync)
+  const shouldSyncSrc = chromiumNeedsUpdate || syncWithForce ||
+      program.sync_src || program.cleanup
```

Comment:
Are we sure we should always sync chromium if the user passes the `--cleanup` flag? I believe they will use this flag if something is going wrong. e.g. our scripts are complaining because an old dep exists on the fs within src/brave but it has missing license information. However, it will be a surprise if chromium also gets updated because this flag was included.
If chromium doesn't need update, then I don't think they will have a problem that requires `-D` to be run for chromium...
We can also log a warning to let the user know that since chromium does not need an update, then `--cleanup` (or whatever we end up calling it) will not be run in that directory, and they can force it with `--force`

Reply:
Also a nit: would be good to log a message if the user sees the message 
that chromium sync not needed but we still do it anyway because sync_src was specified as true.

Reply:
> Are we sure we should always sync chromium if the user passes the --cleanup flag? I believe they will use this flag if something is going wrong. e.g. our scripts are complaining because an old dep exists on the fs within src/brave but it has missing license information.

agree. it was also unclear to me if we should do the sync if `cleanup` was passed. let's not do it.

Reply:
> Are we sure we should always sync chromium if the user passes the `--cleanup` flag?

`--sync_chromium_and_delete_unused_deps` should be pretty clear that it will force the sync and will do a cleanup.